libraries/AP_EFI/AP_EFI_EMUECU.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ 
+#include <AP_HAL/AP_HAL.h>
+#include "AP_EFI_EMUECU.h"
+#include <unistd.h>
+
+#if EFI_ENABLED
+#include <AP_SerialManager/AP_SerialManager.h>
+
+extern const AP_HAL::HAL &hal;
+
+AP_EFI_EMUECU::AP_EFI_EMUECU(AP_EFI &_frontend):
+    AP_EFI_Backend(_frontend)
+{
+    port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_EFI, 0);
+}
+
+
+void AP_EFI_EMUECU::update()
+{
+    if (!port) {
+        return;
+    }
+
+    uint32_t n = port->available();
+    n = MIN(n, 1200U);
+    if (n == 0) {
+        return;
+    }
+
+    while (n) {
+        uint8_t b = port->read();
+        linebuf[line_len++] = b;
+        if (line_len == sizeof(linebuf)-1) {
+            // overflow, reset
+            line_len = 0;
+            break;
+        }
+        if (b == '\n') {
+            // got a full line
+            process_line();
+            line_len = 0;
+        }
+    }
+}
+
+/*
+  process one line of JSON
+ */
+void AP_EFI_EMUECU::process_line(void)
+{
+    for (uint16_t i=0; i<ARRAY_SIZE(keytable); i++) {
+        const struct keytable &key = keytable[i];
+
+        // find key
+        const char *p = strstr(linebuf, key.key);
```

Comment:
A little uneasy on guarantee of null-termination of linebuf.  More reading might make me happy.

Reply:
the sizeof(linebuf)-1 guarantees it